Regression since "[X86][AVX10] Permit AVX512 options/features used together with AVX10 (#71318)"

Since f229ba4e8d6bb24622b2d41b3c89585cb9005d8d, "[X86][AVX10] Permit AVX512 options/features used together with AVX10 (#71318)", some uses/configurations of x86 intrinsics fail. This happens if trying to configure Qt for x86_64, where the Qt cmake configuration fails with the following error:

```
ERROR: x86 intrinsics support missing. Check your compiler settings. If this is an
error, report at https://bugreports.qt.io with your compiler ID and version,
and this output: 

${TEST_x86intrin_OUTPUT} 
```
This is a regression, as builds of Clang prior to f229ba4e8d6bb24622b2d41b3c89585cb9005d8d can build the same version of Qt just fine.

The failing test for x86 intrinsics can be reduced down to this:
```c
#include <immintrin.h>   
int main(int argc, char **argv) {
    __m256i ptrs = _mm256_maskz_loadu_epi64(0, argv);
    return 0; 
}
```
```console
$ clang -target x86_64-linux-gnu -march=cannonlake -c intrin-regression.c
In file included from intrin-regression.c:1:
In file included from /home/martin/clang-nightly-sat/lib/clang/18/include/immintrin.h:109:
/home/martin/clang-nightly-sat/lib/clang/18/include/avx512vlintrin.h:5552:18: error: always_inline function '_mm256_setzero_si256' requires target feature 'evex512', but would be inlined into function '_mm256_maskz_loadu_epi64' that is compiled without support for 'evex512'
 5552 |                  _mm256_setzero_si256 (),
      |                  ^
1 error generated.
```

CC @phoebewang

--- a/clang/lib/Basic/Targets/X86.cpp
+++ b/clang/lib/Basic/Targets/X86.cpp
@@ -122,7 +122,7 @@ bool X86TargetInfo::initFeatureMap(
   std::vector<std::string> UpdatedAVX10FeaturesVec;
   enum { FE_NOSET = -1, FE_FALSE, FE_TRUE };
   int HasEVEX512 = FE_NOSET;
-  bool HasAVX512F = Features.count("avx512f") && Features["avx512f"];
+  bool HasAVX512F = Features.lookup("avx512f");
   bool HasAVX10 = false;
   bool HasAVX10_512 = false;
   std::string LastAVX10;

--- a/clang/lib/Basic/Targets/X86.cpp
+++ b/clang/lib/Basic/Targets/X86.cpp
@@ -122,7 +122,7 @@ bool X86TargetInfo::initFeatureMap(
   std::vector<std::string> UpdatedAVX10FeaturesVec;
   enum { FE_NOSET = -1, FE_FALSE, FE_TRUE };
   int HasEVEX512 = FE_NOSET;
-  bool HasAVX512F = false;
+  bool HasAVX512F = Features["avx512f"];
   bool HasAVX10 = false;
   bool HasAVX10_512 = false;
   std::string LastAVX10;

--- a/clang/lib/Basic/Targets/X86.cpp
+++ b/clang/lib/Basic/Targets/X86.cpp
@@ -122,7 +122,7 @@ bool X86TargetInfo::initFeatureMap(
   std::vector<std::string> UpdatedAVX10FeaturesVec;
   enum { FE_NOSET = -1, FE_FALSE, FE_TRUE };
   int HasEVEX512 = FE_NOSET;
-  bool HasAVX512F = Features["avx512f"];
+  bool HasAVX512F = Features.count("avx512f") && Features["avx512f"];
   bool HasAVX10 = false;
   bool HasAVX10_512 = false;
   std::string LastAVX10;

--- a/clang/lib/Basic/Targets/X86.cpp
+++ b/clang/lib/Basic/Targets/X86.cpp
@@ -122,7 +122,7 @@ bool X86TargetInfo::initFeatureMap(
   std::vector<std::string> UpdatedAVX10FeaturesVec;
   enum { FE_NOSET = -1, FE_FALSE, FE_TRUE };
   int HasEVEX512 = FE_NOSET;
-  bool HasAVX512F = Features["avx512f"];
+  bool HasAVX512F = false;
   bool HasAVX10 = false;
   bool HasAVX10_512 = false;
   std::string LastAVX10;
@@ -180,6 +180,8 @@ bool X86TargetInfo::initFeatureMap(
       Diags.Report(diag::warn_invalid_feature_combination)
           << LastAVX10 + (HasEVEX512 == FE_TRUE ? " +evex512" : " -evex512");
     UpdatedFeaturesVec.push_back(HasAVX10_512 ? "+evex512" : "-evex512");
+  } else if (HasEVEX512 == FE_FALSE) {
+    UpdatedFeaturesVec.push_back("-evex512");
   }
 
   if (!TargetInfo::initFeatureMap(Features, Diags, CPU, UpdatedFeaturesVec))

--- a/clang/lib/Basic/Targets/X86.cpp
+++ b/clang/lib/Basic/Targets/X86.cpp
@@ -122,7 +122,7 @@ bool X86TargetInfo::initFeatureMap(
   std::vector<std::string> UpdatedAVX10FeaturesVec;
   enum { FE_NOSET = -1, FE_FALSE, FE_TRUE };
   int HasEVEX512 = FE_NOSET;
-  bool HasAVX512F = false;
+  bool HasAVX512F = Features["avx512f"];
   bool HasAVX10 = false;
   bool HasAVX10_512 = false;
   std::string LastAVX10;
@@ -180,8 +180,6 @@ bool X86TargetInfo::initFeatureMap(
       Diags.Report(diag::warn_invalid_feature_combination)
           << LastAVX10 + (HasEVEX512 == FE_TRUE ? " +evex512" : " -evex512");
     UpdatedFeaturesVec.push_back(HasAVX10_512 ? "+evex512" : "-evex512");
-  } else if (HasEVEX512 == FE_FALSE) {
-    UpdatedFeaturesVec.push_back("-evex512");
   }
 
   if (!TargetInfo::initFeatureMap(Features, Diags, CPU, UpdatedFeaturesVec))

--- a/clang/lib/Basic/Targets/X86.cpp
+++ b/clang/lib/Basic/Targets/X86.cpp
@@ -123,8 +123,8 @@ bool X86TargetInfo::initFeatureMap(
   enum { FE_NOSET = -1, FE_FALSE, FE_TRUE };
   int HasEVEX512 = FE_NOSET;
   bool HasAVX512F = Features["avx512f"];
-  bool HasAVX10 = false;
-  bool HasAVX10_512 = false;
+  bool HasAVX10 = Features["avx10.1-256"];
+  bool HasAVX10_512 = Features["avx10.1-512"];
   std::string LastAVX10;
   std::string LastAVX512;
   for (const auto &Feature : FeaturesVec) {

--- a/clang/lib/Basic/Targets/X86.cpp
+++ b/clang/lib/Basic/Targets/X86.cpp
@@ -123,8 +123,8 @@ bool X86TargetInfo::initFeatureMap(
   enum { FE_NOSET = -1, FE_FALSE, FE_TRUE };
   int HasEVEX512 = FE_NOSET;
   bool HasAVX512F = Features["avx512f"];
-  bool HasAVX10 = Features["avx10.1-256"];
-  bool HasAVX10_512 = Features["avx10.1-512"];
+  bool HasAVX10 = false;
+  bool HasAVX10_512 = false;
   std::string LastAVX10;
   std::string LastAVX512;
   for (const auto &Feature : FeaturesVec) {

--- a/clang/lib/Basic/Targets/X86.cpp
+++ b/clang/lib/Basic/Targets/X86.cpp
@@ -123,8 +123,8 @@ bool X86TargetInfo::initFeatureMap(
   enum { FE_NOSET = -1, FE_FALSE, FE_TRUE };
   int HasEVEX512 = FE_NOSET;
   bool HasAVX512F = Features.lookup("avx512f");
-  bool HasAVX10 = false;
-  bool HasAVX10_512 = false;
+  bool HasAVX10 = Features.lookup("avx10.1-256");
+  bool HasAVX10_512 = Features.lookup("avx10.1-512");
   std::string LastAVX10;
   std::string LastAVX512;
   for (const auto &Feature : FeaturesVec) {

--- a/clang/test/CodeGen/X86/pr72106.c
+++ b/clang/test/CodeGen/X86/pr72106.c
@@ -1,0 +1,10 @@
+// RUN: %clang_cc1 -ffreestanding -target-cpu cannonlake -emit-llvm < %s | FileCheck %s
+
+#include <immintrin.h>
+
+int main(int argc, char **argv) {
+  // CHECK-LABEL: @main
+  // CHECK: @llvm.masked.load.v4i64.p0
+  __m256i ptrs = _mm256_maskz_loadu_epi64(0, argv);
+  return 0;
+}

